### PR TITLE
remove usage of XFormInstance.DELETED (part 1)

### PR DESCRIPTION
**Pull request by minhaminha**
## Technical Summary

Changes that stemmed from [this comment](https://github.com/dimagi/commcare-hq/pull/33057#discussion_r1219686338) from [this PR](https://github.com/dimagi/commcare-hq/pull/33057) which was blocking that PR from moving forward. 

This part is removing the usage of XFormInstance.DELETED state. Following these [guidelines for updating sql fucntions](https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/README.md) (that were making use of that state), those functions have been udpated in a duplicate file, with the removal of the old files/functions to come in a followup PR. Because of that, I can't actually remove the state outright in this PR since those old functions still reference them.

### Safety story

I'm not sure how to test this beyond having existing tests pass and passing review.

### QA Plan

n/a

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by snopoke** [comment]
In terms of updating the SQL functions, there are docs about that process here: https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/README.md


> * In [soft_delete_forms2.sql](https://github.com/dimagi/commcare-hq/blob/master/corehq/sql_accessors/sql_templates/soft_delete_forms2.sql), assuming that `state = state | {{ FORM_STATE_DELETED }}` is more or less just a regular `or` statement, should state just be None if deleted_on has a value instead? 

This is actually a bitwise OR. Since the states are integers with non-overlapping bits (powers of 2) we can use bitwise operations to set / check the state. This allows a form to have multiple states e.g. `NORMAL && DELETED`.

I think the 'deleted' state was the only reason we needed this since the other states are mutually exclusive.

In the deletion functions we can remove any changes to the `state` since the status will now be based on the presence of the `deleted_on` field.

Same goes for [this line](https://github.com/dimagi/commcare-hq/blob/f738c4c5315174c88ddc2f84053b81e6f9e397e4/corehq/form_processor/models/forms.py#L644) in `soft_delete`.

This line can be replaced by setting the `deleted_on` date to the current date. The purpose of this line is to make the in-memory form object match the DB state (since the DB state is being manipulated directly).


> * In [soft_undelete_forms2.sql](https://github.com/dimagi/commcare-hq/blob/master/corehq/sql_accessors/sql_templates/soft_undelete_forms2.sql), the line `state = state & ~{{ FORM_STATE_DELETED }}` I assume is meant to read as state = state unless its state = DELETED but what exactly should state be here if DELETED is being removed in favor of using deleted_on? (Basically same question as above but wanted to make sure my understanding of this line was correct)

Same as above, you can remove the line that manipulates the value of `state`.

File: corehq/apps/data_pipeline_audit/management/commands/find_sql_forms_not_in_es.py
Comment (snopoke) [inline corehq/apps/data_pipeline_audit/management/commands/find_sql_forms_not_in_es.py:71]: I think this can be simplified to `Q(state__in=[XFormInstance.NORMAL, XFormInstance.ARCHIVED])`

File: corehq/form_processor/models/forms.py
Comment (snopoke) [inline corehq/form_processor/models/forms.py:191]: This line can be removed entirely

File: corehq/sql_accessors/migrations/0055_set_form_modified_on.py
Comment (snopoke) [inline corehq/sql_accessors/migrations/0055_set_form_modified_on.py:9]: this will have to stay until the SQL template files which reference it are removed (since you'll be creating new copies of the functions initially)

File: corehq/form_processor/migrations/0094_auto_20230823_1600.py
Comment (millerdev) [inline corehq/form_processor/migrations/0094_auto_20230823_1600.py:1]: I know other existing migrations in this app are not setting a great example, but it would be nice to give this migration a meaningful name.

**Comment by minhaminha** [comment]
Thanks so much @snopoke for your detailed response! Since I can't remove the context in the migration files until after the sql file updates are done I've made another [PR here](https://github.com/dimagi/commcare-hq/pull/33397) to add the new, updated sql functions, which will be followed by the old function deletion PR after that's merged.

**Comment by minhaminha** [comment]
Backtracking a little... I re-added the deleted state for now so that I can remove it alongside the old sql templates in another PR and added the changes made in the other PR (now closed), mostly so I can be sure the tests pass.

**Review by snopoke — APPROVED** [review]

File: corehq/form_processor/migrations/0094_add_partial_index_xforms.py
Comment (snopoke) [inline corehq/form_processor/migrations/0094_add_partial_index_xforms.py:16]: Just thinking about this some more - won't this lock the table while the index is being created?

You may need to take a more manual approach to creating the index so that you can make it a 'concurrent' operation. Here's an example: https://github.com/dimagi/commcare-hq/blob/0bae5ccf834ce0bc97d3e36470a05f4049665b0a/corehq/apps/sms/migrations/0045_auto_20200902_0938.py#L15

It's probably also a good idea to split the index addition into a separate PR so that it can be rolled out independent of the other changes (since it may require running outside of a deploy - at least on prod).

Comment (minhaminha) [reply corehq/form_processor/migrations/0094_add_partial_index_xforms.py:16]: thanks for the example Simon! and good call in separating these out... I probably should've realized this needed to be two PRs when I put that '+' in the PR title

**Comment by minhaminha** [comment]
Rebased (as best I could) to remove references to the partial index migration.

**Review by millerdev — APPROVED** [review]

**Comment by minhaminha** [comment]
> Does the index creation PR need to be applied and merged before this one? I assume the index is being added to make the functions in this PR more efficient.

I didn't intend for it to be ordered like that but that's a great point. I'll hold off on merging until the other one's approved too.

**Pull request merged**
